File: IITD/Codemania_Classical_1.0/D. Deserted Freshers/sol-dsu.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <utility>
#include <vector>

using namespace std;

struct dsu {
    int n;
    vector<int32_t> par, siz;

    dsu(int n) : n(n), par(n), siz(n) {
        for (int i = 0; i < n; ++i) {
            make_set(i);
        }
    };

    inline void make_set(int32_t v) {
        par[v] = v;
        siz[v] = 1;
    }

    int32_t find_set(int32_t v) {
        if (v == par[v]) return v;
        return par[v] = find_set(par[v]);
    }

    void union_sets(int32_t a, int32_t b) {
        a = find_set(a);
        b = find_set(b);
        if (a != b) {
            if (siz[a] < siz[b]) swap(a, b);
            par[b] = a;
            siz[a] += siz[b];
        }
    }

    vector<vector<int>> components() {
        vector<int> representative(n), component_size(n);
        for (int i = 0; i < n; ++i) {
            representative[i] = find_set(i);
            component_size[representative[i]]++;
        }
        vector<vector<int>> components(n);
        for (int i = 0; i < n; ++i) components[i].reserve(component_size[i]);
        for (int i = 0; i < n; ++i) components[representative[i]].push_back(i);
        components.erase(
            remove_if(components.begin(), components.end(),
                      [](const vector<int>& v) { return v.empty(); }),
            components.end());
        return components;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<pair<int, int>> a(n);
    for (int i = 0; i < n; ++i) {
        cin >> a[i].first;
        a[i].second = i;
    }
    sort(a.rbegin(), a.rend());

    dsu country(n);

    for (int i = 0, u, v; i < m; ++i) {
        cin >> u >> v;
        country.union_sets(u - 1, v - 1);
    }

    auto components = move(country.components());

    vector<int> indices(n);
    iota(indices.begin(), indices.end(), 0);
    sort(indices.begin(), indices.end(), [&](const int i, const int j) {
        return components[i].size() > components[j].size();
    });

    vector<int> ans(n, -1);
    int current_rank = 0;
    for (auto index : indices) {
        for (auto i : components[index]) {
            ans[a[current_rank].second] = i + 1;
            current_rank++;
        }
    }

    for (auto x : ans) cout << x << ' ';
    cout << '\n';

    return 0;
}

File: IITD/Codemania_Classical_1.0/D. Deserted Freshers/sol-dsu.py
# Fast IO

import sys
input = sys.stdin.readline
def print(x, end='\n'):
    sys.stdout.write(str(x) + end)

# IO helpers

def get_int():
    return int(input())
def get_list_ints():
    return list(map(int, input().split()))
def get_char_list():
    s = input()
    return list(s[:len(s) - 1])
def get_tuple_ints():
    return tuple(map(int, input().split()))
def print_iterable(p):
    print(" ".join(map(str, p)))

class DSU:
    def __init__(self, n):
        self.n = n
        self.par = [i for i in range(n)]
        self.siz = [1 for i in range(n)]

    def make_set(self, v):
        self.par[v] = v
        self.siz[v] = 1

    def find_set(self, v):
        if v == self.par[v]:
            return v
        self.par[v] = self.find_set(self.par[v])
        return self.par[v]

    def union_sets(self, a, b):
        a = self.find_set(a)
        b = self.find_set(b)
        if a != b:
            if self.siz[a] < self.siz[b]:
                a, b = b, a
            self.par[b] = a
            self.siz[a] += self.siz[b]

    def components(self):
        representative = [self.find_set(i) for i in range(self.n)]
        answer = [[] for i in range(self.n)]
        for i in range(self.n):
            answer[representative[i]].append(i)
        return [x for x in answer if x]

def main():

    n, m = get_tuple_ints()
    a = get_list_ints()

    visited = [False for i in range(n)]
    g = [[] for i in range(n)]
    components = [[] for i in range(n)]

    dsu = DSU(n)

    for i in range(m):
        u, v = get_tuple_ints()
        dsu.union_sets(u - 1, v - 1)

    components = dsu.components()

    indices = sorted([i for i in range(n)], key=lambda x : -a[x])
    components.sort(key=lambda x : -len(x))

    current_rank = 0
    assigned_vertex = [0 for i in range(n)]
    for component in components:
        for x in component:
            assigned_vertex[indices[current_rank]] = x + 1
            current_rank += 1

    print_iterable(assigned_vertex)
    pass

if __name__ == '__main__':
    main()
